### PR TITLE
Assortment of improvements for CentOS 7

### PR DESCRIPTION
**Pull request by gdhgdhgdh**
Hi,

Here are a handful of changes that I've had to make during my use of your module. No longer does it complain that I need to recreate the Ceph rbd datastore each time :)

Cheers,
Gavin.

### CODE REVIEW

**Comment by rwaffen** [comment]
hi there, i will have a closer look at it tomorrow. Generally looks good for me, but @tuxmea might have some comments for the types/providers.

**Comment by gdhgdhgdh** [comment]
Cool, thanks. 

On the failed tests - I don't understand why it failed the Debian APT key tests, however the RedHat `/sbin/brctl` symlink makes sense. Commit 23d9c8e specifically removed the symlink for RedHat since otherwise you end up with a symlink to itself and no usable `brctl` binary :(

**Comment by rwaffen** [comment]
i don't see the point with brctl... this works in centos 6. the brctl package delivers only a binary in /usr/sbin but opennebula needs it in /sbin. do you use centos 7, was it changed there?

```
[rwaffen@srv ~]$ rpm -ql bridge-utils
/usr/sbin/brctl
/usr/share/doc/bridge-utils-1.2
/usr/share/doc/bridge-utils-1.2/AUTHORS
/usr/share/doc/bridge-utils-1.2/COPYING
/usr/share/doc/bridge-utils-1.2/FAQ
/usr/share/doc/bridge-utils-1.2/HOWTO
/usr/share/man/man8/brctl.8.gz
```

File: manifests/params.pp [hunk 1]
```diff
@@ -141,7 +147,7 @@
       $libvirtd_srv = 'libvirtd'
       $libvirtd_cfg = '/etc/sysconfig/libvirtd'
       $libvirtd_source = 'puppet:///modules/one/libvirtd.sysconfig'
-      $rubygems       = ['builder']
+      $rubygems       = ['builder', 'sinatra']
```
Comment (rwaffen) [inline manifests/params.pp]: why do we need sinatra?

File: manifests/prerequisites.pp [hunk 2]
```diff
@@ -31,7 +31,7 @@
             }
         }
         'Debian' : {
-            if ($one::params::one_repo_enable == 'true') {
+            if ($one::params::one_repo_enable == true) {
```
Comment (rwaffen) [inline manifests/prerequisites.pp]: here is the error with the tests... you turned 'true' to true (boolean) ... and it seems puppet can't handle this.
Look at the discussion in #130 ... there are some problems with booleans and hiera values.

File: manifests/compute_node/config.pp [hunk 3]
```diff
@@ -69,13 +69,15 @@
     source => 'puppet:///modules/one/sudoers_imaginator',
   }
 
-  file { '/sbin/brctl':
-    ensure => link,
-    target => '/usr/sbin/brctl',
+  if $::osfamily == 'Debian' {
```
Comment (rwaffen) [inline manifests/compute_node/config.pp]: if we change this ... we should change the tests, too. ;-)

```
  1) one with hiera config on RedHat with one module should contain File[/sbin/brctl] with ensure => "link"
     Failure/Error: it { should contain_file('/sbin/brctl').with_ensure('link') }
       expected that the catalogue would contain File[/sbin/brctl]
```

**Comment by gdhgdhgdh** [comment]
Thanks for taking the time to review and provide feedback :)

I've been using CentOS 7 for this build (my usual choice is Debian and I fancied a change). `bridge-utils` looks similar to CentOS 6:

```
[root@userver1 ~]# rpm -ql bridge-utils
/usr/sbin/brctl
/usr/share/doc/bridge-utils-1.5
/usr/share/doc/bridge-utils-1.5/AUTHORS
/usr/share/doc/bridge-utils-1.5/COPYING
/usr/share/doc/bridge-utils-1.5/FAQ
/usr/share/doc/bridge-utils-1.5/HOWTO
/usr/share/man/man8/brctl.8.gz
```

However the part which may differ are these symlinks in the root:

```
[root@userver1 ~]# ls -l /
total 32
lrwxrwxrwx.   1 root root    7 Mar 20 21:33 bin -> usr/bin
dr-xr-xr-x.   4 root root 4096 Mar 20 22:10 boot
drwxr-xr-x   22 root root 3440 Mar 22 18:26 dev
drwxr-xr-x.  94 root root 8192 Mar 22 16:19 etc
drwxr-xr-x.   2 root root    6 Jun 10  2014 home
lrwxrwxrwx.   1 root root    7 Mar 20 21:33 lib -> usr/lib
lrwxrwxrwx.   1 root root    9 Mar 20 21:33 lib64 -> usr/lib64
drwxr-xr-x.   2 root root    6 Jun 10  2014 media
drwxr-xr-x.   5 root root   37 Mar 22 20:41 mnt
drwxr-xr-x.   2 root root    6 Jun 10  2014 opt
dr-xr-xr-x  147 root root    0 Mar 21 18:19 proc
dr-xr-x---.  11 root root 4096 Mar 22 15:55 root
drwxr-xr-x   32 root root 1040 Mar 22 21:40 run
lrwxrwxrwx.   1 root root    8 Mar 20 21:33 sbin -> usr/sbin
```

Hence when Puppet installs a symlink at `/sbin/brctl` it's actually overwriting the binary at `/usr/sbin/brctl` - the symlinking of `/bin` and `/sbin` are new in RHEL7/CentOS7:

Ref: https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Migration_Planning_Guide/sect-Red_Hat_Enterprise_Linux-Migration_Planning_Guide-File_System_Layout.html#sect-Red_Hat_Enterprise_Linux-Migration_Planning_Guide-File_System_Layout-New_layout_for_root_file_system

In that case, my change is incorrect since it should only apply when `$::lsbmajdistrelease == 7`

File: manifests/params.pp (hunk 1 above)
Comment (gdhgdhgdh) [inline manifests/params.pp]: I found that the `sunstone` web UI refused to start, and threw a Ruby stack trace complaining about being unable to require `sinatra`  - perhaps a packaging bug in upstream's RPM's?

File: manifests/prerequisites.pp (hunk 2 above)
Comment (gdhgdhgdh) [inline manifests/prerequisites.pp]: Ahh that's it - when I saw the lint tests throw dozens of WARNINGs and a couple of ERRORs, I ran `puppet-lint -f` over all the `.pp` manifests. It looks like `puppet-lint` was a little over-zealous in its optimisation. Thanks for the explainer.

File: manifests/compute_node/config.pp (hunk 3 above)
Comment (gdhgdhgdh) [inline manifests/compute_node/config.pp]: I can't really argue with that :) It might finally be time that I learned how to write rspec tests...

**Comment by gdhgdhgdh** [comment]
Interesting. My changes have somehow managed to break compatibility with Ruby 1.8.7.

Aside from why this might be the case, is that version of Ruby still considered important to support? Upstream abandoned it in 2013, and even Ubuntu's extended support stopped in July 2014.

**Comment by tuxmea** [comment]
Many people still use puppet on ruby 1.8
The error is within the version compare.
You are comparing string (facter) with integer.
Please make use of the versioncmp function.
Please change your code to:
```
diff --git a/manifests/compute_node/config.pp b/manifests/compute_node/config.pp
index 913fb65..636ae7b 100644
--- a/manifests/compute_node/config.pp
+++ b/manifests/compute_node/config.pp
@@ -76,7 +76,7 @@ class one::compute_node::config (
     }
   }

-  if $::osfamily == 'RedHat' and $::lsbmajdistrelease < 7  {
+  if $::osfamily == 'RedHat' and versioncmp($::lsbmajdistrelease, '7') < 0  {
     file { '/sbin/brctl':
       ensure => link,
       target => '/usr/sbin/brctl',
```
Then the tests will pass.

**Comment by gdhgdhgdh** [comment]
Thanks for the insight - I'm still fairly new with Puppet so I would never have realised the `versioncmp` method. 

Was there something specific in the build logs which pointed to that conclusion, or simply greater experience and exposure?

**Comment by tuxmea** [comment]
When running the spec tests one saw the error message: "undefined method `<' for :undef:Symbol"
Something like this is always due to comparing incompatible data.

File: manifests/oned/sunstone/config.pp [hunk 4]
```diff
@@ -20,6 +20,7 @@
   $listen_ip          = $one::params::sunstone_listen_ip,
   $enable_support     = $one::params::enable_support,
   $enable_marketplace = $one::params::enable_marketplace,
+  $tmpdir             = $one::params::sunstone_tmpdir,
```
Comment (tuxmea) [inline manifests/oned/sunstone/config.pp]: Please add the parameter tmpdir also to manifests/params.pp and set it to OpenNebula default (/var/tmp).

**Comment by rwaffen** [comment]
tests are green now :smile: ... looks good to me... @tuxmea are we ready to merge?

**Comment by tuxmea** [comment]
No. Merge conflicts! @gdhgdhgdh you need to rebase your branch to upstream master.

**Comment by rwaffen** [comment]
ah, okay, i see. there's also a big warning left to my (now grey) merge button.

**Comment by gdhgdhgdh** [comment]
Tonight was not the right time to try a rebase or merge for the first time.

**Comment by tuxmea** [comment]
;-)

**Comment by rwaffen** [comment]
so, merged... lets see what happen :smile:

**Pull request merged**